crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs

```diff
@@ -0,0 +1,303 @@
+    type Codomain = CodomainShape<E::G1>;
+    type Domain = (E::ScalarField, Vec<E::ScalarField>);
+
+    fn apply(&self, input: &Self::Domain) -> Self::Codomain {
+        self.apply_msm(self.msm_terms(input))
+    }
+}
+
+impl<'a, E: Pairing> fixed_base_msms::Trait for Homomorphism<'a, E> {
+    type Base = E::G1Affine;
+    type CodomainShape<T>
+        = CodomainShape<T>
+    where
+        T: CanonicalSerialize + CanonicalDeserialize + Clone;
+    type MsmInput = fixed_base_msms::MsmInput<Self::Base, Self::Scalar>;
+    type MsmOutput = E::G1;
+    type Scalar = E::ScalarField;
+
+    fn msm_terms(&self, input: &Self::Domain) -> Self::CodomainShape<Self::MsmInput> {
+        assert!(
+            self.lagr_g1.len() >= input.1.len(),
+            "Not enough Lagrange basis elements for univariate hiding KZG: required {}, got {}",
+            input.1.len(),
+            self.lagr_g1.len()
+        );
+
+        let mut scalars = Vec::with_capacity(input.1.len() + 1);
+        scalars.push(input.0);
+        scalars.extend_from_slice(&input.1);
+
+        let mut bases = Vec::with_capacity(input.1.len() + 1);
+        bases.push(self.xi_1);
+        bases.extend(&self.lagr_g1[..input.1.len()]);
+
+        CodomainShape(fixed_base_msms::MsmInput { bases, scalars })
+    }
+
+    fn msm_eval(bases: &[Self::Base], scalars: &[Self::Scalar]) -> Self::MsmOutput {
+        E::G1::msm(bases, &scalars).expect("MSM computation failed in univariate KZG")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_ec::pairing::Pairing;
+    use ark_poly::{univariate::DensePolynomial, Polynomial};
+    use ark_std::{rand::thread_rng, UniformRand};
+
+    // TODO: Should set up a PCS trait, then make these tests generic?
+    fn generic_test_open_and_verify_roundtrip<E: Pairing>() {
+        let mut rng = thread_rng();
+        let group_data = GroupGenerators::sample(&mut rng);
+
+        type Fr<E> = <E as Pairing>::ScalarField;
+
+        let m = 64;
+        let xi = Fr::<E>::rand(&mut rng);
+        let tau = Fr::<E>::rand(&mut rng);
+        let (vk, ck) = setup::<E, _>(m, group_data, Trapdoor { xi, tau }, &mut rng);
+
+        let f_coeffs: Vec<Fr<E>> = (0..m).map(|_| Fr::<E>::rand(&mut rng)).collect();
+        let poly = DensePolynomial::<Fr<E>> { coeffs: f_coeffs };
+
+        // Polynomial values at the roots of unity
+        let f_evals: Vec<Fr<E>> = ck
+            .roots_of_unity_in_eval_dom
+            .iter()
+            .map(|&gamma| poly.evaluate(&gamma))
+            .collect();
+
+        let rho = CommitmentRandomness::<E>(Fr::<E>::rand(&mut rng));
+        let s = CommitmentRandomness::<E>(Fr::<E>::rand(&mut rng));
+        let x = Fr::<E>::rand(&mut rng);
+        let y =
+            polynomials::barycentric_eval(&f_evals, &ck.roots_of_unity_in_eval_dom, x, ck.m_inv);
+
+        // Commit to f
+        let comm = super::commit_with_randomness(&ck, &f_evals, &rho);
+
+        // Open at x
```

Comment:
This will fail when x is a root of unity.

Negligible in our large prime field so we are good 👌